python/lsst/ts/MTAirCompressor/aircompressor_csc.py

```diff
@@ -178,51 +169,55 @@ def run_sim():
                 None, run_sim
             )
             await asyncio.sleep(2)
-            self.client = ModbusClient(host="localhost", port=5020)
-        else:
-            self.client = ModbusClient(host=self.hostname)
-
-        if self.client.connect() is False:
-            self.log.error(f"Cannot establish connection to {self.hostname}")
 
-        self.telemetry_task = asyncio.create_task(self.telemetry_loop())
+        try:
+            await self.connect()
+            if self.poll_task.done():
+                self.poll_task = asyncio.create_task(self.poll_loop())
+        except ModbusError as er:
+            await self.log_modbus_error(er, "Starting up:", True)
+            return
 
     async def end_enable(self, data):
         """Power on compressor after switching to enable state. Raise exception
-        if compressor cannot be powered on."""
-        poweredOn = self.client.write_register(
-            Registers.REMOTE_CMD, 0xFF01, unit=self.unit
-        )
-        if poweredOn.isError():
-            raise ModbusError(
-                "Cannot power on compressor", poweredOn, Registers.REMOTE_CMD
+        if compressor cannot be powered on. Ignore state transition triggered
+        by auto update."""
+        if self._status_update is True:
+            return
+        if self._start_by_remote is False:
+            raise RuntimeError(
+                "Compressor isn't in remote mode - cannot be powered on remotely"
             )
 
-    async def end_disable(self, data):
-        """Power off compressor after switching to disable state."""
-        poweredDown = self.client.write_register(
-            Registers.REMOTE_CMD, 0xFF00, unit=self.unit
-        )
-        if poweredDown.isError():
-            raise ModbusError(
-                "Cannot power off compressor", poweredDown, Registers.REMOTE_CMD
-            )
-        self.telemetry_task.cancel()
+        try:
+            self.model.power_on(unit=self.unit)
+        except ModbusError as ex:
+            await self.log_modbus_error(ex, "Cannot power on compressor")
+
+    async def begin_disable(self, data):
+        """Power off compressor before switching to disable state. Ignore state
+        transition triggered by auto update."""
+        if self._status_update is True:
+            return
+        try:
+            self.model.power_off(self.unit)
+        except ModbusError as ex:
+            try:
+                if ex.exception.original_code & 0x10 == 0x10:
+                    raise RuntimeError(
+                        "Compressor isn't in remote mode - cannot be powered off"
+                    )
```

Comment:
I'm a bit surprised you try to turn it off even if you already know the cmpressor is not in remote mode. But I guess it makes sense if the mode can be changed while you are using it?

Reply:
https://jira.lsstcorp.org/browse/DM-35280 for future improvements.